### ISSUE TITLE
Regent: CI regression triggered by LuaJIT upgrade

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1279]
I'm tracking a CI regression in Regent. Symptoms look like `assertion failed: mismatch in data left over in future`, as seen in CI jobs [here](https://gitlab.com/StanfordLegion/legion/-/jobs/2593834529) and [here](https://gitlab.com/StanfordLegion/legion/-/jobs/2599409362). This is not a part of the code that has changed recently in either Regent or Legion. The CI regression is new (~days).

Initial investigation seems to show that the regression is sensitive to a recent LuaJIT upgrade in Terra. [This branch](https://github.com/StanfordLegion/terra/tree/legion-check-luajit) of Terra fails, immediately after the LuaJIT upgrade. [This branch](https://github.com/StanfordLegion/terra/tree/legion-check-luajit-old), which is identical but with LuaJIT put back to the old version, passes. I ran the CI job 4 times to be sure. (The failure is non-deterministic and seems to reproduce about 50% of the time.)

I have been able to reproduce on a Sapling compute node with the command line: `PATH="/scratch2/eslaught/ccache/install_bin:$PATH" LAUNCHER="mpirun -n 2 -H localhost -oversubscribe" REALM_SYNTHETIC_CORE_MAP= DEBUG=0 REALM_NETWORKS=gasnetex CONDUIT=mpi GASNET=$PWD/gasnet/release USE_CUDA=1 TEST_CUDA=1 NO_PRETTY=1 ./travis.py` (using ccache to make rebuilding repeatedly slightly less painful).

I'm currently doing a bisect through LuaJIT history to see if I can pinpoint the problem.

While I'm not done with the bisect yet, my guess (based on looking at the remaining 8 commits or so) is that it's likely to be this one:

```
commit a44f53acf53603e7d9b88352de035b1804be4e88 (HEAD)
Author: Mike Pall <mike>
Date:   Mon Jun 15 12:21:05 2020 +0200

    Use a securely seeded global PRNG for the VM.
    
    It's not 2005 anymore.
```

If this commit means that hash table iteration order is non-deterministic, that would indeed cause issues for Regent. We rely on hash table iteration to deterministic (at least with reasonable key types like integers and strings) in order to maintain deterministic compilation. If this is violated, it's going to be quite easy to miscompile Regent code in distributed tests.

**Comment by elliottslaughter** [comment]
Nope, I'm wrong. The commit in my last comment is actually fine---I let it run up to 17 times to be sure. Instead the bad commit is the one immediately after:

```
ff34b48ddd6f2b3bdd26d6088662a214ba6b0288 is the first bad commit
commit ff34b48ddd6f2b3bdd26d6088662a214ba6b0288
Author: Mike Pall <mike>
Date:   Tue Jun 23 03:06:45 2020 +0200

    Redesign and harden string interning.
    
    Up to 40% faster on hash-intensive benchmarks.
    With some ideas from Sokolov Yura.
```

I suppose that, depending on the exact nature of what changed in string interning, my initial diagnosis could still be correct. Anything that results in nondeterminism in the iteration order would be a problem, and I use string keys extensively as "safe" hashable values in `data.map` and elsewhere, so it's likely this could be the issue.

**Comment by elliottslaughter** [comment]
Pushed a tentative fix in 9db92cc2787649cbe56be0ac38f80e6bbf10bc18, will watch the CI and do more testing to see if it solves the original problem.

**Comment by elliottslaughter** [comment]
I have a new fix in a branch that reduces the probability of occurrence by at least 10&times;. RDIR is currently misbehaving on one test case, otherwise it just needs a bit more cleanup and verification before I can merge. My original diagnosis appears to be correct; it just seems to be a matter of tracking down the remaining locations where issues can occur to make sure we're completely deterministic.

**Comment by elliottslaughter** [comment]
I pushed what is hopefully a permanent fix in baac365b610f4a971be460d937c272bad9dcd3ed and will leave the test to run for a while on sapling to make sure it's really resolved.

**Comment by elliottslaughter** [comment]
I ran 236 successful iterations on sapling, so I'm going to consider this resolved.
